Add static analysis unit test for Controller classes

RyanM-RMA opened the pull request:
Related to #1349, I'll be adding a unit test for static analysis of of the Controller implementations.

rma-rripken committed:
Trying to do some static analysis of controller classes.

RyanM-RMA committed:
Add JavaParser dependency for static analysis of Controller impls

RyanM-RMA committed:
Removing previous parser dependency

RyanM-RMA committed:
Adding a more generic and faster method for identifying classes under a specific package that inherits from a provided type.

RyanM-RMA committed:
Refactor OpenApiTestHelperTest to separate Handler and CrudHandler tests, add UsersController validation.

RyanM-RMA committed:
Refactor OpenApiTestHelper and related tests to use separated OpenApiDocInfo for improved parameter handling, and add tests for interface-based handlers.

RyanM-RMA committed:
Refactor OpenApiDocInfo and related classes to improve type handling, introduce OpenApiDocTestInfo, and update tests accordingly. Add new OpenApiDocTest for handler and CRUD handler documentation validation.

RyanM-RMA committed:
Attempting to resolve build failure

RyanM-RMA committed:
Refactor OpenApiTestHelper and OpenApiDocTest to introduce JavaParser-based static analysis for improved validation and modularized helper methods.

RyanM-RMA committed:
Refactor OpenApiDocInfo, OpenApiTestHelper, and related tests to introduce OpenApiParamInfo for enhanced parameter metadata handling.  Introducing a quick test for TimeSeriesController for faster turnaround.

RyanM-RMA committed:
Added testing for ignored OpenApi methods, need to talk with Corps about consistency with how the ignored methods are handled.

RyanM-RMA committed:
Introduce `OpenApiParamUsage` for enhanced parameter tracking, refactor related methods for improved query and path parameter parsing.

RyanM-RMA committed:
Refactor parameter handling in OpenApi classes: replace `Set` with `List` for parameter collections so it includes order, add null-handling tracking, and enhance test coverage for query and path parameter validation.

RyanM-RMA committed:
Adding better handling for queryParamAsClass methods, resolving "?" parameter issues and using set instead of list for parameter usage

RyanM-RMA committed:
Update `OpenApiTestHelperTest` to assert parameters using `OpenApiParamInfo` with constants from `Controllers`.

rma-rripken commented on cwms-data-api/src/test/java/cwms/cda/api/OpenApiDocTest.java:
```diff
@@ -0,0 +1,486 @@
+        String missingInfo = missingItems.stream()
+                                         .map(OpenApiParamInfo::getName)
+                                         .collect(Collectors.joining(", "));
+        assertAll(() -> assertTrue(receivedItems.isEmpty(), "Found extra path parameters: " + extraInfo),
```
I feel like this should explain what "extra" means.

RyanM-RMA replied:
Updated it to say it's a used undocumented path parameter.

rma-rripken commented on cwms-data-api/src/test/java/cwms/cda/api/OpenApiDocTest.java:
```diff
@@ -0,0 +1,486 @@
+                                         .map(OpenApiParamInfo::getName)
+                                         .collect(Collectors.joining(", "));
+        assertAll(() -> assertTrue(receivedItems.isEmpty(), "Found extra path parameters: " + extraInfo),
+                  () -> assertTrue(missingItems.isEmpty(), "Found missing path parameters: " + missingInfo),
```
Also here, what does "missing" mean?

RyanM-RMA replied:
Updated to say that it's a documented path parameter that's not used

rma-rripken commented on cwms-data-api/src/test/java/cwms/cda/api/OpenApiDocTest.java:
```diff
@@ -0,0 +1,486 @@
+        List<OpenApiParamUsageInfo> verifiedUsages = new ArrayList<>();
+        List<OpenApiParamInfo> expectedParams = new ArrayList<>();
+        List<OpenApiParamInfo> missingItems = new ArrayList<>();
+        List<OpenApiParamUsageInfo> receivedItems = new ArrayList<>(receivedQueryParameters);
```
These List seem like they should be Set

RyanM-RMA replied:
Updated to Sets, let me know if I missed any

rma-rripken commented on cwms-data-api/src/test/java/cwms/cda/api/OpenApiDocTest.java:
```diff
@@ -0,0 +1,486 @@
+            // This is effectively like...we have a name, but we're not sure where it's coming from and we're
+            // Not a string literal.
+            try {
+                Field field = Controllers.class.getField(arg.asNameExpr().getNameAsString());
```
This is using Reflection to resolve the reference not JavaParser.  Thats interesting.   And its assuming the declaring class is always Controllers.   I think you could do something like this and know what the declaring class is.    Its usually going to be Controllers or the current class but it would be better to know than to guess.  

} else if (arg instanceof NameExpr) {
            NameExpr ne = (NameExpr) arg;
            ResolvedValueDeclaration resolved = ne.resolve();
            String declaringClass = resolved.asField().declaringType().getQualifiedName();
            String fieldName = arg.asNameExpr().getNameAsString();

RyanM-RMA replied:
This was a bit of a wild ride to get right, but I got the symbol resolver stuff figured out.  It does make the tests slower, but not by a significant amount.

rma-rripken commented on cwms-data-api/src/test/java/cwms/cda/api/OpenApiDocTest.java:
```diff
@@ -0,0 +1,486 @@
+    void test_time_series_controller() throws IOException {
+        OpenApiDocTestInfo testInfo = OpenApiTestHelper.readOpenApiDocs(CrudHandler.class, TimeSeriesController.class);
+        CompilationUnit compilationUnit = OpenApiTestHelper.readCompilationUnit(testInfo.getClazz());
+        assertAll(buildTestAssertions(compilationUnit, testInfo));
+    }
+
+    private Stream<Executable> buildTestAssertions(CompilationUnit compilationUnit, OpenApiDocTestInfo testInfo) {
+        return testInfo.getMethodDocs()
+                       .stream()
+                       .map(docInfo -> validateOpenApiDoc(compilationUnit, docInfo, testInfo.getClazz()));
+    }
+
+    private Executable validateOpenApiDoc(CompilationUnit unit, OpenApiDocInfo testInfo, Class<?> clazz){
+        Executable output;
+        if (testInfo.isIgnored()) {
+            output = testIgnoredMethod(unit, testInfo, clazz);
+        } else {
+            OpenApiParamUsage parsedParamInfo = parseParamInfo(unit, clazz, testInfo.getMethod());
+            output = testMethod(testInfo, parsedParamInfo);
+        }
+        return output;
+    }
+
+    private Executable testIgnoredMethod(CompilationUnit unit, OpenApiDocInfo testInfo, Class<?> clazz) {
+
+        // 16/24 cases throw an UnsupportedOperationException, so verify that behavior here for now.
+        // Other cases use the context to provide additional feedback, but the feedback is not consistent.
+        // Could be an improvement for the corps to define it more strictly here.
+        MethodDeclaration method = getMethodDeclaration(unit, testInfo.getMethod());
+        boolean throwsException = method.findAll(ThrowStmt.class)
+                                        .stream()
+                                        .anyMatch(throwStmt -> {
+                                            if (throwStmt.getExpression() instanceof ObjectCreationExpr) {
+                                                ObjectCreationExpr creation = (ObjectCreationExpr) throwStmt.getExpression();
+                                                return creation.getType().getNameAsString().equals("UnsupportedOperationException");
+                                            }
+                                            return false;
+                                        });
+        return () -> assertTrue(throwsException, clazz.getSimpleName() + "::" + testInfo.getMethod().getName() + " is marked as ignored, but does not throw an UnsupportedOperationException.");
+    }
```
I'm not convinced that we need to look for UnsupportedOperationException.  Or, I'm not convinced that is the only "right" way for something to be unimplemented.    There's at least one controller that does: 
```
 ctx.status(HttpCode.NOT_IMPLEMENTED).json(CdaError.notImplemented());
```
And that seems pretty clear and clean to me.  I'd be curious what other patterns are in use besides those two.

MikeNeilson replied:
One could argue, in light of that, that we should determine a single right way for us.

RyanM-RMA replied:
I updated this to look for `ctx.status(HttpServletResponse.SC_NOT_IMPLEMENTED).json(CdaError.notImplemented())`

rma-rripken commented:
This ended up needing more code from JavaParser than I anticipated from the beginning.  I don't know JavaParser well enough to know if there are shortcuts we are missing but from what I've seen in the results it is catching the errors I was hoping we'd find.

RyanM-RMA committed:
Adding symbol resolver to more simply resolve symbols in JavaParser

Clearing up messaging for documented unused parameters vs undocumented used parameters.
Adding code for resolving references and fields and such down to their actual values.

rma-rripken reviewed (dismissed):
We should check with Mike on whether we submit this enabled or not?  Maybe a @Tag on the test class so that its submitted but not run during the automated build?

MikeNeilson commented:
> We should check with Mike on whether we submit this enabled or not? Maybe a @tag on the test class so that its submitted but not run during the automated build?

Gentleman, I shall scream it again... `EnableIfProperty`! and then add another job to the build so we see the results but it doesn't block the builds yet.

RyanM-RMA commented:
> > We should check with Mike on whether we submit this enabled or not? Maybe a @tag on the test class so that its submitted but not run during the automated build?
> 
> Gentleman, I shall scream it again... `EnableIfProperty`! and then add another job to the build so we see the results but it doesn't block the builds yet.

Oooh, I don't believe I have used this, thank you for the quick feedback @MikeNeilson.  Give me a few minutes, and I'll get that going.

MikeNeilson commented:
> > > We should check with Mike on whether we submit this enabled or not? Maybe a @tag on the test class so that its submitted but not run during the automated build?
> > 
> > 
> > Gentleman, I shall scream it again... `EnableIfProperty`! and then add another job to the build so we see the results but it doesn't block the builds yet.
> 
> Oooh, I don't believe I have used this, thank you for the quick feedback @MikeNeilson. Give me a few minutes, and I'll get that going.

FYI did the name for memory, it might be slightly different than that exact text.

RyanM-RMA committed:
Adding EnabledIfEnvironmentVariable to the OpenApiDocTest and adding a new job to run the static analysis separately.

MikeNeilson commented on .github/workflows/build.yml:
```diff
@@ -56,6 +56,31 @@ jobs:
+      - name: build and test
+        id: thebuild
+        run: ./gradlew :cwms-data-api:test --tests "cwms.cda.api.OpenApiDocTest" --info --init-script init.gradle
+        continue-on-error: true
```
remove this and set the below to "always"

We've tweaked the repo to have specific "required builds" this one for now is informational, but we still want it to show up as "failed" to draw eyes.

RyanM-RMA replied:
I think I updated this right, let me know.

MikeNeilson replied:
Yep, though looks like you need to update the branch, below is still showing the update button.

RyanM-RMA committed:
Update OpenApi test workflow to always publish JUnit reports regardless of test outcome
